File: src/lib/classes/Track.ts
import type { Album } from "./Album";
import { Artist } from "./Artist";
import { ExternalId } from "./ExternalId";
import { ExternalUrl } from "./ExternalUrl";
import { TrackLink } from "./TrackLink";

export class Track implements SpotifyData {
  // Saved Track 
  added_at?: string;

  // Track Full
  album?: Album;
  external_ids?: ExternalId;
  popularity?: number;
  
  // Track Simplified
  artists!: Artist[];
  available_markets?: string[];
  disc_number!: number;
  duration_ms!: number;
  explicit!: boolean;
  external_urls!: ExternalUrl;
  href!: string;
  id!: string;
  is_playable?: boolean;
  linked_from?: TrackLink;
  name!: string;
  preview_url!: string;
  track_number!: number;
  type!: 'track';
  uri!: string;

  private constructor() {
  }


  public static fromSavedTrack(saved: SpotifyApi.SavedTrackObject): Track {
    const track = Track.fromFullTrack(saved.track);
    track.added_at = saved.added_at;
    return track;
  }

  public static fromFullTrack(full: SpotifyApi.TrackObjectFull): Track {
    const track = Track.fromSimplifiedTrack(full);
    track.artists = full.artists.map(a => Artist.fromSimplifiedArtist(a));
    track.external_ids = ExternalId.fromExternalIdObject(full.external_ids);
    track.popularity = full.popularity;
    return track;
  }

  public static fromSimplifiedTrack(simplified: SpotifyApi.TrackObjectSimplified): Track {
    const track = new Track();
    track.artists = simplified.artists.map(a => Artist.fromSimplifiedArtist(a));
    track.available_markets = simplified.available_markets;
    track.disc_number = simplified.disc_number;
    track.duration_ms = simplified.duration_ms;
    track.explicit = simplified.explicit;
    track.external_urls = ExternalUrl.fromExternalUrlObject(simplified.external_urls);
    track.href = simplified.href;
    track.id = simplified.id;
    track.is_playable = simplified.is_playable;
    track.linked_from = simplified.linked_from && TrackLink.fromTrackLinkObject(simplified.linked_from);
    track.name = simplified.name;
    track.preview_url = simplified.preview_url;
    track.track_number = simplified.track_number;
    track.type = simplified.type;
    track.uri = simplified.uri;
    return track;
  }
}

File: src/main/java/de/wittenbude/expandify/models/spotifydata/Track.java
package de.wittenbude.expandify.models.spotifydata;

import com.neovisionaries.i18n.CountryCode;
import de.wittenbude.expandify.models.spotifydata.helper.TrackLink;
import lombok.*;
import org.springframework.data.annotation.Id;
import org.springframework.data.mongodb.core.mapping.Document;
import org.springframework.data.mongodb.core.mapping.DocumentReference;
import se.michaelthelin.spotify.enums.ModelObjectType;

import java.util.Arrays;
import java.util.List;
import java.util.Map;

@Data
@NoArgsConstructor
@Document(collection = "tracks")
public class Track {
    @DocumentReference(lazy = true)
    private AlbumSimplified album;
    @DocumentReference(lazy = true)
    private List<ArtistSimplified> artists;
    private CountryCode[] availableMarkets;
    private Integer discNumber;
    private Integer durationMs;
    private Boolean explicit;
    private Map<String, String> externalIds;
    private Map<String, String> externalUrls;
    private String href;
    @Id
    private String id;
    private Boolean isPlayable;
    private TrackLink linkedFrom;
    private String restrictionReason;
    private String name;
    private Integer popularity;
    private String previewUrl;
    private Integer trackNumber;
    private ModelObjectType type;
    private String uri;


    public Track(se.michaelthelin.spotify.model_objects.specification.Track track) {
        this.album = new AlbumSimplified(track.getAlbum());
        this.artists = Arrays.stream(track.getArtists()).map(ArtistSimplified::new).toList();
        this.discNumber = track.getDiscNumber();
        this.explicit = track.getIsExplicit();
        this.externalIds = track.getExternalIds().getExternalIds();
        this.externalUrls = track.getExternalUrls().getExternalUrls();
        this.href = track.getHref();
        this.id = track.getId();
        this.isPlayable = track.getIsPlayable();
        this.linkedFrom = new TrackLink(track.getLinkedFrom());
        this.restrictionReason = track.getRestrictions().getReason();
        this.name = track.getName();
        this.popularity = track.getPopularity();
        this.previewUrl = track.getPreviewUrl();
        this.trackNumber = track.getTrackNumber();
        this.type = track.getType();
        this.uri = track.getUri();
    }
}
